### ISSUE TITLE
Support Ecto-style ownership mechanism

### ISSUE THREAD

**Issue by hubertlepicki** [plataformatec/mox#5]
Here's something I think would be pretty cool, is to allow mocking with Wallaby sessions.

I propose a similar solution that is used with Ecto, where we have a plug, that has to be configured in endpoint.ex for test env. 

Then in a `setup` block we would have to switch mode to shared:

    Mox.mode({:shared, self()})

and all the mocks we set under the tests would be available.

Behind the scenes this would work similar to Ecto sandbox. We probably need to use different default header than `user-agent`, however, since this one seems already taken by Ecto. I think it should be custom header anyway, even in Ecto, if the major browser drivers support it, but that's whole other story.

What do you think? I could possibly put in some time to get this going.

**Comment by josevalim** [comment]
@hubertlepicki yes, we can definitely do this, I was just too lazy to do it. :) I think the first step is to remove the registry and put it behind a GenServer - keeping the same feature set. Then we can discuss how to add some sharing options. What do you think? Let me know if i can be of any guidance.

**Comment by brandonjoyce** [comment]
If I understand this correctly, this would allow us to test separate processes that collaborate with our mock?  Something like this?
```
# example GenServer that gets the calc module injected.
defmodule CalcServer do
  use GenServer
  def start_link(calc_mod) do
    GenServer.start_link(__MODULE__, calc_mod)
  end

  def add(pid, x, y) do
     GenServer.call(pid, {:add, x, y})
  end

  def handle_call({:add, x, y}, calc_mod) do
    {:reply, calc_mod.add(x, y), calc_mod}
  end
end

# Tests look something like this
Mox.mode(CalcMock, {:manual, self()})
{:ok, pid} = CalcGenServer.start_link(CalcMock)
CalcMock
|> expect(:add, fn(x, y) -> x + y end)
|> allow(pid, self()) 

CalcServer.add(pid, 1, 1)  # uses setups from above because the pid was allowed. 
```

This would be fantastic!

**Comment by josevalim** [comment]
@brandonjoyce yes, precisely.

**Comment by hubertlepicki** [comment]
Yes, I think the above makes total sense. I was thinking about integration testing with real browser and something like Wallaby, but it is in fact the same case when you want to write mocks/stubs for the processes.

I'll start working on this today, let's see where this leads.

**Comment by hubertlepicki** [comment]
@josevalim I thik I'll need some guidance in the end. I started coding this but uncertain if it's the good way forward.

Basically I think we need to start a GenServer per test process. I think this does not have to be done explicitly, i.e. the first attempt to mock something, or attempt to set mode would start such GenServer up.

 I think we keep this GenServer on hand, can be registered in a Registry.

I do *not* think we need to supervise it / restart it in case there's a bug in mox (we should prob crash the testing process when our genserver crashes), but I was thinking of using supervisor we have and :simple_one_for_one to do the actual spawning of these GenServers.

Does above remotely aligns with what you were thinking?

**Comment by josevalim** [comment]
We don't need a GenServer per test. Since the goal is to collaborate between multiple processes, we need to have a central entity.

@hubertlepicki did you try replacing the Registry by a GenServer as a starting point? Simply get rid of the Registry altogether and query the GenServer, as if the GenServer was the registry. If you do this migration first, it will probably become clearer what is the next step.

**Comment by hubertlepicki** [comment]
OK makes sense. I had a feeling I've overthinked the above solution so good that I asked.

**Comment by dcaixinha** [comment]
Hey guys, amazing work with this! 💯 This is definitely a pain point I had while testing Elixir, and up to this point I had been maintaining my own `GenServer`s on mock modules in order test thing that aren't in the current process.

Given that @josevalim stated that replacing the `Registry` with a `GenServer` (implemented in #8) was the first step to achieve this, what do you guys think is missing, and with what can I help with?

**Comment by hubertlepicki** [comment]
Hey,

I think the next step is to split the state in two maps, as also discussed
and allow other processes to access mocks as the last comment from #8
states. This is not yet started by me. If you want to do it now go ahead,
otherwise I will pick it up at some point tonight (CEST).

The step after that will involve working on a Plug that would allow mocks
to be used by processes used by Cowboy/Phoenix.
